File: web/lib/transactionFilters.ts
// Pure, cross-platform transaction filtering. Mirrored in Swift
// (iOS/.../TransactionFilters.swift) and locked by shared golden vectors
// (shared/test-vectors/transaction-filters.json) so the two clients can't drift.
// See specs/006-transaction-filters.
import type { Transaction, TransactionCategory } from './types'

export interface FilterCriteria {
  query: string
  categories: TransactionCategory[]
  kind: 'all' | 'expense' | 'income'
  sources: string[]
  owners: string[]
  /** Inclusive ISO lower bound (`date >= from`). */
  dateFrom: string | null
  /** Exclusive ISO upper bound (`date < to`). Half-open `[from, to)`. */
  dateTo: string | null
}

export interface FilterContext {
  /** ownerId → display name, for search-by-owner-name. */
  ownerNames: Record<string, string>
}

export function emptyCriteria(): FilterCriteria {
  return { query: '', categories: [], kind: 'all', sources: [], owners: [], dateFrom: null, dateTo: null }
}

function matchesQuery(tx: Transaction, q: string, ownerNames: Record<string, string>): boolean {
  if (q === '') return true
  if (tx.merchant.toLowerCase().includes(q)) return true
  if (tx.source.toLowerCase().includes(q)) return true
  if (tx.category.toLowerCase().includes(q)) return true
  for (const id of tx.owner_ids) {
    const name = ownerNames[id]
    if (name && name.toLowerCase().includes(q)) return true
  }
  return false
}

/** Subset of `txs` passing ALL dimensions (OR within a multi-select). Order preserved. */
export function filterTransactions(txs: Transaction[], c: FilterCriteria, ctx: FilterContext): Transaction[] {
  const q = c.query.trim().toLowerCase()
  const fromT = c.dateFrom ? new Date(c.dateFrom).getTime() : null
  const toT = c.dateTo ? new Date(c.dateTo).getTime() : null
  return txs.filter((tx) => {
    if (!matchesQuery(tx, q, ctx.ownerNames)) return false
    if (c.categories.length > 0 && !c.categories.includes(tx.category)) return false
    if (c.kind !== 'all' && tx.kind !== c.kind) return false
    if (c.sources.length > 0 && !c.sources.includes(tx.source)) return false
    if (c.owners.length > 0 && !tx.owner_ids.some((id) => c.owners.includes(id))) return false
    if (fromT !== null || toT !== null) {
      const t = new Date(tx.date).getTime()
      if (fromT !== null && t < fromT) return false
      if (toT !== null && t >= toT) return false
    }
    return true
  })
}

/** Count of non-default dimensions (for the "N filters active" badge). */
export function activeFilterCount(c: FilterCriteria): number {
  let n = 0
  if (c.query.trim() !== '') n++
  if (c.categories.length > 0) n++
  if (c.kind !== 'all') n++
  if (c.sources.length > 0) n++
  if (c.owners.length > 0) n++
  if (c.dateFrom !== null || c.dateTo !== null) n++
  return n
}

/** Distinct non-empty sources present in the transactions, alphabetized. */
export function availableSources(txs: Transaction[]): string[] {
  const set = new Set<string>()
  for (const tx of txs) {
    const s = tx.source.trim()
    if (s) set.add(s)
  }
  return [...set].sort((a, b) => a.localeCompare(b))
}

const pad = (n: number) => String(n).padStart(2, '0')

/** "2026-05" → half-open UTC month window. Timezone-stable. */
export function monthBounds(yyyymm: string): { dateFrom: string; dateTo: string } {
  const m = yyyymm.match(/^(\d{4})-(\d{2})$/)
  if (!m) throw new Error(`INVALID_MONTH: ${JSON.stringify(yyyymm)}`)
  const y = Number(m[1])
  const mo = Number(m[2])
  if (mo < 1 || mo > 12) throw new Error(`INVALID_MONTH: ${yyyymm}`)
  const ny = mo === 12 ? y + 1 : y
  const nm = mo === 12 ? 1 : mo + 1
  return { dateFrom: `${y}-${pad(mo)}-01T00:00:00.000Z`, dateTo: `${ny}-${pad(nm)}-01T00:00:00.000Z` }
}

File: iOS/Ortho-iOS/Features/Transactions/TransactionFilters.swift
import Foundation

// Pure, cross-platform transaction filtering — mirror of web/lib/transactionFilters.ts.
// Locked by the shared golden vectors (shared/test-vectors/transaction-filters.json);
// see Ortho-iOSTests/TransactionFilterParityTests.swift. See specs/006-transaction-filters.

struct FilterCriteria: Equatable {
    enum Kind: String, CaseIterable { case all, expense, income, transfer }

    var query: String = ""
    var categories: Set<TransactionCategory> = []
    var kind: Kind = .all
    var sources: Set<String> = []
    var owners: Set<User.ID> = []
    /// Inclusive lower bound (`date >= from`).
    var dateFrom: Date? = nil
    /// Exclusive upper bound (`date < to`). Half-open `[from, to)`.
    var dateTo: Date? = nil

    static let empty = FilterCriteria()
}

struct FilterContext {
    /// ownerId → display name, for search-by-owner-name.
    var ownerNames: [User.ID: String]
}

/// Subset of `txs` passing ALL dimensions (OR within a multi-select). Order preserved.
func filterTransactions(_ txs: [Transaction], _ c: FilterCriteria, _ ctx: FilterContext) -> [Transaction] {
    // `.whitespacesAndNewlines` matches JS `String.prototype.trim()` (which also
    // strips newlines) so a trailing "\n" in the query behaves identically.
    let q = c.query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    return txs.filter { tx in
        // search
        if !q.isEmpty {
            var hit = tx.merchant.lowercased().contains(q)
                || tx.source.lowercased().contains(q)
                || tx.category.rawValue.lowercased().contains(q)
            if !hit {
                for id in tx.ownerIDs {
                    if let name = ctx.ownerNames[id], name.lowercased().contains(q) { hit = true; break }
                }
            }
            if !hit { return false }
        }
        // category
        if !c.categories.isEmpty && !c.categories.contains(tx.category) { return false }
        // kind
        if c.kind != .all && tx.kind.rawValue != c.kind.rawValue { return false }
        // source
        if !c.sources.isEmpty && !c.sources.contains(tx.source) { return false }
        // owner
        if !c.owners.isEmpty && tx.ownerIDs.isDisjoint(with: c.owners) { return false }
        // date (half-open)
        if let from = c.dateFrom, tx.date < from { return false }
        if let to = c.dateTo, tx.date >= to { return false }
        return true
    }
}

/// Count of non-default dimensions (for the "N filters active" badge).
func activeFilterCount(_ c: FilterCriteria) -> Int {
    var n = 0
    if !c.query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { n += 1 }
    if !c.categories.isEmpty { n += 1 }
    if c.kind != .all { n += 1 }
    if !c.sources.isEmpty { n += 1 }
    if !c.owners.isEmpty { n += 1 }
    if c.dateFrom != nil || c.dateTo != nil { n += 1 }
    return n
}

/// Distinct non-empty sources present in the transactions, alphabetized.
func availableSources(_ txs: [Transaction]) -> [String] {
    var set = Set<String>()
    for tx in txs {
        let s = tx.source.trimmingCharacters(in: .whitespacesAndNewlines)
        if !s.isEmpty { set.insert(s) }
    }
    // Localized, case-insensitive order — mirrors web `a.localeCompare(b)` (a
    // locale-aware collation), NOT Swift's default Unicode code-point `sorted()`.
    return set.sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }
}

/// "2026-05" → half-open UTC month window. Timezone-stable (matches the TS).
///
/// Input is validated STRICTLY against `^\d{4}-\d{2}$`: "2026-5", "26-5",
/// "2026-005", and "2026--5" are all rejected (exactly like web's regex).
/// Web THROWS `INVALID_MONTH` on a bad string; every Swift call site here uses
/// `if let (from, to) = monthBounds(...)`, so the mirrored strictness surfaces
/// as the documented safe failure `nil` (a bad month yields no window) rather
/// than a trap.
func monthBounds(_ yyyymm: String) -> (dateFrom: Date, dateTo: Date)? {
    // Strict shape: 4 ASCII digits, '-', 2 ASCII digits. `split` is too lenient
    // (it drops empty fields), so validate the exact layout character by character.
    let chars = Array(yyyymm)
    func isDigit(_ c: Character) -> Bool { c.isASCII && c.isNumber }
    guard chars.count == 7, chars[4] == "-",
          isDigit(chars[0]), isDigit(chars[1]), isDigit(chars[2]), isDigit(chars[3]),
          isDigit(chars[5]), isDigit(chars[6]),
          let y = Int(String(chars[0...3])),
          let mo = Int(String(chars[5...6])),
          (1...12).contains(mo) else { return nil }
    var cal = Calendar(identifier: .gregorian)
    cal.timeZone = TimeZone(identifier: "UTC")!
    guard let from = cal.date(from: DateComponents(year: y, month: mo, day: 1)),
          let to = cal.date(byAdding: .month, value: 1, to: from) else { return nil }
    return (from, to)
}
